python/vision_explanation_methods/evaluation/pointing_game.py

```diff
@@ -0,0 +1,163 @@
+
+        detections = self._model.predict(img_input)
+
+        saliency_scores = drise.DRISE_saliency(
+                model=self._model,
+                # Repeated the tensor to test batching
+                image_tensor=img_input,
+                target_detections=detections,
+                # This is how many masks to run -
+                # more is slower but gives higher quality mask.
+                number_of_masks=num_masks,
+                mask_padding=None,
+                device=self._device,
+                # This is the resolution of the random masks.
+                # High resolutions will give finer masks, but more need to be
+                # run.
+                mask_res=(2, 2),
+                verbose=True  # Turns progress bar on/off.
+            )
+
+        temp = saliency_scores[0][index]['detection']
+
+        temp[temp < threshold] = -1
+        return temp
+
+    def visualize_highly_salient_pixels(self,
+                                        img,
+                                        saliency_scores):
+        """
+        Create figure of highly salient pixels.
+
+        :param img: PIL test image
+        :type img: PIL.Image
+        :param saliency_scores: 2D matrix representing the saliency scores
+            of each pixel in an image
+        :type saliency_scores: List[Tensor]
+        :return: Overlay of the saliency scores on top of the image
+        :rtype: Figure
+        """
+        fig, ax = pl.subplots(1, 1, figsize=(10, 10))
+
+        viz.visualize_image_attr(
+            np.transpose(
+                saliency_scores.detach().cpu().numpy(),
+                (1, 2, 0)),
+            np.transpose(T.ToTensor()(img).detach().cpu().numpy(), (1, 2, 0)),
+            method="blended_heat_map",
+            sign="positive",
+            show_colorbar=True,
+            cmap=pl.cm.get_cmap("Blues"),
+            title="Pointing Game Visualization",
+            plt_fig_axis=(fig, ax),
+            use_pyplot=False
+        )
+
+        return fig
+
+    def calculate_gt_salient_pixel_overlap(self,
+                                           saliency_scores: List[Tensor],
+                                           gt_bbox: List):
+        """
+        Calculate percent of overlap between salient pixels and gt bbox.
+
+        :param saliency_scores: 2D matrix representing the saliency scores
+            of each pixel in an image
+        :type saliency_scores: List[Tensor]
+        :param gt_bbox: bounding box for ground truth prediction
+        :type gt_bbox: List
+        :return: return percent of salient pixel overlap with the ground truth
+        :rtype: Float
+        """
+        saliency_scores = torch.tensor(saliency_scores)
+        gt_bbox = torch.tensor(gt_bbox)
+
+        gt_mask = torch.zeros_like(saliency_scores, dtype=torch.bool)
+        gt_mask[gt_bbox[0]:gt_bbox[2], gt_bbox[1]:gt_bbox[3]] = True
+
+        positive_mask = saliency_scores > 0
```

Comment:
Nit: Presume this can also be done in torch, dk if that may be faster under the hood

Reply:
The latency difference between using torch.gt() and creating a boolean mask using a comparison operator (>) will depend on various factors, including the size of the input tensors and the hardware on which the computations are performed.

In general, torch.gt() and the comparison operator will have similar performance characteristics because torch.gt() is implemented as a low-level operation using efficient tensor operations. However, the exact latency difference can vary based on the specific hardware, memory layout, and other factors.

It's worth noting that creating a boolean mask using a comparison operator (>) may involve additional memory allocation for the resulting boolean tensor, which can slightly impact the overall latency compared to using torch.gt() directly.

In practice, if you are working with large tensors or performing computationally intensive operations, the performance difference between torch.gt() and creating a boolean mask using a comparison operator is likely to be negligible. It is recommended to choose the approach that aligns better with the overall code structure and readability rather than solely focusing on potential latency differences.


Source: ChatGPT :)

Reply:
If I understand correctly, using torch.gt can be better in terms of memory of the entire, though may be by a small margin in practice. Another benefit could be that the entire metric could be calculated on a GPU rather than combining intermediate computations from GPU & CPU.

May encourage the below update just coz it's a trivial update into a principally better logic.

```suggestion
        positive_mask = torch.gt(saliency_scores, 0)
```

Reply:
If I understand correctly, torch.gt may be slightly better in terms of memory of the tensor, though perhaps in practice the difference could still be minimal. Another benefit could be the entire metric would be calculated on GPU instead of combining intermediate computations from CPU & GPU.

Would encourage the below update as it's a trivial update into a principally better logic
```suggestion
        positive_mask = torch.gt(saliency_scores, 0)
```

Reply:
Sure, my interpretation was that the difference was negligible as size increased and hardware changed, but your point about offering faster computation for GPU is valid. Will commit the change.